ckanext/canada/harvesters.py

```diff
@@ -0,0 +1,616 @@
+        portal_ini = plugins.toolkit.config.get('ckanext.canada.portal_sync.portal_ini')
+        if not registry_ini:
+            raise Exception('ckanext.canada.portal_sync.registry_ini not defined')
+        if not isfile(registry_ini):
+            raise Exception('Cannot find file %s, defined from ckanext.canada.portal_sync.registry_ini' % registry_ini)
+        if not portal_ini:
+            raise Exception('ckanext.canada.portal_sync.portal_ini not defined')
+        if not isfile(portal_ini):
+            raise Exception('Cannot find file %s, defined from ckanext.canada.portal_sync.portal_ini' % portal_ini)
+
+        registry_config = ConfigParser()
+        registry_config.read(registry_ini)
+        try:
+            registry_datastore_url = registry_config.get('app:main', 'ckan.datastore.write_url')
+        except NoOptionError:
+            raise Exception('Cannot find ckan.datastore.write_url in app:main section of Registry INI')
+        self.registry_datastore_url = str(_get_engine_from_url(registry_datastore_url).url)  # ensure that this is actually a DS database
+        portal_config = ConfigParser()
+        portal_config.read(portal_ini)
+        try:
+            portal_datastore_url = portal_config.get('app:main', 'ckan.datastore.write_url')
+        except NoOptionError:
+            raise Exception('Cannot find ckan.datastore.write_url in app:main section of Portal INI')
+        self.portal_datastore_url = str(_get_engine_from_url(portal_datastore_url).url)  # ensure that this is actually a DS database
+
+        #FIXME: figure out how to get the data from and to the Portal in fetch_stage and import_stage
+        self.registry = LocalCKAN()
+        self.portal = LocalCKAN()
+
+
+    @classmethod
+    def last_error_free_job(cls, harvest_job):
+        """Look for jobs with no gather errors"""
+        jobs = (model.Session.query(HarvestJob)
+                .filter(HarvestJob.source == harvest_job.source)
+                .filter(HarvestJob.gather_started != None)  # noqa: E711
+                .filter(HarvestJob.status == 'Finished')
+                .filter(HarvestJob.id != harvest_job.id)
+                .filter(
+            ~exists().where(
+                HarvestGatherError.harvest_job_id == HarvestJob.id))
+                .outerjoin(HarvestObject,
+                           and_(HarvestObject.harvest_job_id == HarvestJob.id,
+                                HarvestObject.current == False,  # noqa: E712
+                                HarvestObject.report_status != 'not modified'))
+                .options(contains_eager(HarvestJob.objects))
+                .order_by(HarvestJob.gather_started.desc()))
+        # now check them until we find one with no fetch/import errors
+        # if objects count is 0, job was error free
+        for job in jobs:
+            if len(job.objects) == 0:
+                return job
```

Comment:
Do this part in the SQL so that you don't have postgres sending you all the records to iterate over in python, which could be really slow.

Reply:
Class method was copied from the base class of the harvest core code. Not sure how/where to add in the `.filter(HarvestJob.objects == 0)`